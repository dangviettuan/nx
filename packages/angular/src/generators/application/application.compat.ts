import { convertNxGenerator } from '@nrwl/devkit';
import { warnForSchematicUsage } from '../utils/warn-for-schematic-usage';
import application from './application';

export const applicationSchematic = warnForSchematicUsage(
  convertNxGenerator(application)
);
