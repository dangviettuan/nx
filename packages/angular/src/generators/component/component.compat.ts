import componentGenerator from './component';
import { warnForSchematicUsage } from '../utils/warn-for-schematic-usage';
import { convertNxGenerator } from '@nrwl/devkit';

export default warnForSchematicUsage(convertNxGenerator(componentGenerator));
