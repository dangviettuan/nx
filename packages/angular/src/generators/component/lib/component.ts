import type { Tree } from '@nrwl/devkit';
import { logger, readProjectConfiguration, stripIndents } from '@nrwl/devkit';
import { getComponentFileInfo } from '../../utils/file-info';
import { locateLibraryEntryPointFromDirectory } from '../../utils/entry-point';
import { getRelativeImportToFile } from '../../utils/path';
import type { NormalizedSchema } from '../schema';
import { shouldExportInEntryPoint } from './entry-point';
import { findModuleFromOptions } from './module';

export function exportComponentInEntryPoint(
  tree: Tree,
  schema: NormalizedSchema
): void {
  if (!schema.export || (schema.skipImport && !schema.standalone)) {
    return;
  }

  const { root, projectType } = readProjectConfiguration(tree, schema.project);

  if (projectType === 'application') {
    return;
  }

  const { directory, filePath } = getComponentFileInfo(tree, schema);

  const entryPointPath = locateLibraryEntryPointFromDirectory(
    tree,
    directory,
    root,
    schema.projectSourceRoot
  );
  if (!entryPointPath) {
    logger.warn(
      `Unable to determine whether the component should be exported in the library entry point file. ` +
        `The library's entry point file could not be found. Skipping exporting the component in the entry point file.`
    );

    return;
  }

  if (!schema.standalone) {
    const modulePath = findModuleFromOptions(tree, schema, root);
    if (!shouldExportInEntryPoint(tree, entryPointPath, modulePath)) {
      return;
    }
  }

  const relativePathFromEntryPoint = getRelativeImportToFile(
    entryPointPath,
    filePath
  );
  const updateEntryPointContent = stripIndents`${tree.read(
    entryPointPath,
    'utf-8'
  )}
    export * from "${relativePathFromEntryPoint}";`;

  tree.write(entryPointPath, updateEntryPointContent);
}
