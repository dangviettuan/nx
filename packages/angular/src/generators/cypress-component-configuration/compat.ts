import { convertNxGenerator } from '@nrwl/devkit';
import { cypressComponentConfiguration } from './cypress-component-configuration';

export default convertNxGenerator(cypressComponentConfiguration);
