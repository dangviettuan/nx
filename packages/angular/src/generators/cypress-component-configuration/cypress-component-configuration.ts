import { cypressComponentProject } from '@nrwl/cypress';
import { findBuildConfig } from '@nrwl/cypress/src/utils/find-target-options';
import {
  formatFiles,
  generateFiles,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { relative } from 'path';
import { componentTestGenerator } from '../component-test/component-test';
import {
  getComponentsInfo,
  getStandaloneComponentsInfo,
} from '../utils/storybook-ast/component-info';
import { getProjectEntryPoints } from '../utils/storybook-ast/entry-point';
import { getModuleFilePaths } from '../utils/storybook-ast/module-info';
import { CypressComponentConfigSchema } from './schema';

/**
 * This is for cypress built in component testing, if you want to test with
 * storybook + cypress then use the componentCypressGenerator instead.
 */
export async function cypressComponentConfiguration(
  tree: Tree,
  options: CypressComponentConfigSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const installTask = await cypressComponentProject(tree, {
    project: options.project,
    skipFormat: true,
  });

  await updateProjectConfig(tree, options);
  addFiles(tree, projectConfig, options);
  if (options.skipFormat) {
    await formatFiles(tree);
  }
  return () => {
    installTask();
  };
}
function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentConfigSchema
) {
  const cypressConfigPath = joinPathFragments(
    projectConfig.root,
    'cypress.config.ts'
  );

  if (tree.exists(cypressConfigPath)) {
    tree.delete(cypressConfigPath);
  }
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    projectConfig.root,
    {
      tpl: '',
    }
  );

  if (options.generateTests) {
    const entryPoints = getProjectEntryPoints(tree, options.project);

    const componentInfo = [];
    for (const entryPoint of entryPoints) {
      const moduleFilePaths = getModuleFilePaths(tree, entryPoint);
      componentInfo.push(
        ...getComponentsInfo(
          tree,
          entryPoint,
          moduleFilePaths,
          options.project
        ),
        ...getStandaloneComponentsInfo(tree, entryPoint)
      );
    }

    for (const info of componentInfo) {
      if (info === undefined) {
        continue;
      }
      const componentDirFromProjectRoot = relative(
        projectConfig.root,
        joinPathFragments(info.moduleFolderPath, info.path)
      );
      componentTestGenerator(tree, {
        project: options.project,
        componentName: info.name,
        componentDir: componentDirFromProjectRoot,
        componentFileName: info.componentFileName,
        skipFormat: true,
      });
    }
  }
}

async function updateProjectConfig(
  tree: Tree,
  options: CypressComponentConfigSchema
) {
  const found = await findBuildConfig(tree, {
    project: options.project,
    buildTarget: options.buildTarget,
    validExecutorNames: new Set<string>([
      '@nrwl/angular:webpack-browser',
      '@angular-devkit/build-angular:browser',
    ]),
  });

  assertValidConfig(found?.config);

  const projectConfig = readProjectConfiguration(tree, options.project);
  projectConfig.targets['component-test'].options = {
    ...projectConfig.targets['component-test'].options,
    skipServe: true,
    devServerTarget: found.target,
  };
  updateProjectConfiguration(tree, options.project, projectConfig);
}

function assertValidConfig(config: unknown) {
  if (!config) {
    throw new Error(
      'Unable to find a valid build configuration. Try passing in a target for an Angular app. --build-target=<project>:<target>[:<configuration>]'
    );
  }
}

export default cypressComponentConfiguration;
