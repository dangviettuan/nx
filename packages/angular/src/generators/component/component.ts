import type { Tree } from '@nrwl/devkit';
import {
  formatFiles,
  normalizePath,
  readProjectConfiguration,
  readWorkspaceConfiguration,
} from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { pathStartsWith } from '../utils/path';
import { exportComponentInEntryPoint } from './lib/component';
import { normalizeOptions } from './lib/normalize-options';
import type { NormalizedSchema, Schema } from './schema';

export async function componentGenerator(tree: Tree, rawOptions: Schema) {
  const options = normalizeOptions(tree, rawOptions);
  const { projectSourceRoot, ...schematicOptions } = options;

  checkPathUnderProjectRoot(tree, options);

  const angularComponentSchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );
  await angularComponentSchematic(tree, schematicOptions);

  exportComponentInEntryPoint(tree, options);

  await formatFiles(tree);
}

function checkPathUnderProjectRoot(tree: Tree, schema: NormalizedSchema): void {
  if (!schema.path) {
    return;
  }

  const project =
    schema.project ?? readWorkspaceConfiguration(tree).defaultProject;
  const { root } = readProjectConfiguration(tree, project);

  let pathToComponent = normalizePath(schema.path);
  pathToComponent = pathToComponent.startsWith('/')
    ? pathToComponent.slice(1)
    : pathToComponent;

  if (!pathStartsWith(pathToComponent, root)) {
    throw new Error(
      `The path provided for the component (${schema.path}) does not exist under the project root (${root}). ` +
        `Please make sure to provide a path that exists under the project root.`
    );
  }
}

export default componentGenerator;
