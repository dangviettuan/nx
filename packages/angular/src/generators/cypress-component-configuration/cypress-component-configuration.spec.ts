import { installedCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import {
  DependencyType,
  joinPathFragments,
  ProjectGraph,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { applicationGenerator } from '../application/application';
import { componentGenerator } from '../component/component';
import librarySecondaryEntryPointGenerator from '../library-secondary-entry-point/library-secondary-entry-point';
import { libraryGenerator } from '../library/library';
import { cypressComponentConfiguration } from './cypress-component-configuration';

let projectGraph: ProjectGraph;
jest.mock('@nrwl/cypress/src/utils/cypress-version');
jest.mock('@nrwl/devkit', () => ({
  ...jest.requireActual<any>('@nrwl/devkit'),
  createProjectGraphAsync: jest
    .fn()
    .mockImplementation(async () => projectGraph),
}));
describe('Cypress Component Testing Configuration', () => {
  let tree: Tree;
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    tree.write('.gitignore', '');
    mockedInstalledCypressVersion.mockReturnValue(10);
  });

  describe('updateProjectConfig', () => {
    it('should add project config with --target=<project>:<target>', async () => {
      await applicationGenerator(tree, {
        name: 'fancy-app',
      });
      await libraryGenerator(tree, {
        name: 'fancy-lib',
      });
      await componentGenerator(tree, {
        name: 'fancy-cmp',
        project: 'fancy-lib',
        export: true,
      });
      projectGraph = {
        nodes: {
          'fancy-app': {
            name: 'fancy-app',
            type: 'app',
            data: {
              ...readProjectConfiguration(tree, 'fancy-app'),
            },
          },
          'fancy-lib': {
            name: 'fancy-lib',
            type: 'lib',
            data: {
              ...readProjectConfiguration(tree, 'fancy-lib'),
            },
          },
        },
        dependencies: {
          'fancy-app': [
            {
              type: DependencyType.static,
              source: 'fancy-app',
              target: 'fancy-lib',
            },
          ],
        },
      };
      await cypressComponentConfiguration(tree, {
        project: 'fancy-lib',
        buildTarget: 'fancy-app:build',
        generateTests: false,
      });
      expect(
        tree.exists(
          'libs/fancy-lib/src/lib/fancy-cmp/fancy-cmp.component.cy.ts'
        )
      ).toBeFalsy();
      expect(
        readProjectConfiguration(tree, 'fancy-lib').targets['component-test']
      ).toEqual({
        executor: '@nrwl/cypress:cypress',
        options: {
          cypressConfig: 'libs/fancy-lib/cypress.config.ts',
          devServerTarget: 'fancy-app:build',
          skipServe: true,
          testingType: 'component',
        },
      });
    });

    it('should add project config with --target=<project>:<target>:<config>', async () => {
      await applicationGenerator(tree, {
        name: 'fancy-app',
      });
      await libraryGenerator(tree, {
        name: 'fancy-lib',
      });
      await componentGenerator(tree, {
        name: 'fancy-cmp',
        project: 'fancy-lib',
        export: true,
      });
      projectGraph = {
        nodes: {
          'fancy-app': {
            name: 'fancy-app',
            type: 'app',
            data: {
              ...readProjectConfiguration(tree, 'fancy-app'),
            },
          },
          'fancy-lib': {
            name: 'fancy-lib',
            type: 'lib',
            data: {
              ...readProjectConfiguration(tree, 'fancy-lib'),
            },
          },
        },
        dependencies: {
          'fancy-app': [
            {
              type: DependencyType.static,
              source: 'fancy-app',
              target: 'fancy-lib',
            },
          ],
        },
      };
      await cypressComponentConfiguration(tree, {
        project: 'fancy-lib',
        buildTarget: 'fancy-app:build:development',
        generateTests: false,
      });
      expect(
        tree.exists(
          'libs/fancy-lib/src/lib/fancy-cmp/fancy-cmp.component.cy.ts'
        )
      ).toBeFalsy();
      expect(
        readProjectConfiguration(tree, 'fancy-lib').targets['component-test']
      ).toEqual({
        executor: '@nrwl/cypress:cypress',
        options: {
          cypressConfig: 'libs/fancy-lib/cypress.config.ts',
          devServerTarget: 'fancy-app:build:development',
          skipServe: true,
          testingType: 'component',
        },
      });
    });

    it('should throw if --build-target is invalid', async () => {
      await libraryGenerator(tree, {
        name: 'fancy-lib',
      });
      await expect(
        cypressComponentConfiguration(tree, {
          project: 'fancy-lib',
          buildTarget: 'fancy-app:build:development',
          generateTests: false,
        })
      ).rejects
        .toThrow(`Error trying to find build configuration. Try manually specifying the build target with the --build-target flag.
Provided project? fancy-lib
Provided build target? fancy-app:build:development
Provided Executors? @nrwl/angular:webpack-browser, @angular-devkit/build-angular:browser`);
    });
    it('should use own project config', async () => {
      await applicationGenerator(tree, {
        name: 'fancy-app',
      });
      await componentGenerator(tree, {
        name: 'fancy-cmp',
        project: 'fancy-app',
        export: true,
      });
      projectGraph = {
        nodes: {
          'fancy-app': {
            name: 'fancy-app',
            type: 'app',
            data: {
              ...readProjectConfiguration(tree, 'fancy-app'),
            },
          },
        },
        dependencies: {},
      };
      await cypressComponentConfiguration(tree, {
        project: 'fancy-app',
        generateTests: false,
      });
      expect(
        readProjectConfiguration(tree, 'fancy-app').targets['component-test']
      ).toEqual({
        executor: '@nrwl/cypress:cypress',
        options: {
          cypressConfig: 'apps/fancy-app/cypress.config.ts',
          devServerTarget: 'fancy-app:build',
          skipServe: true,
          testingType: 'component',
        },
      });
    });

    it('should use the project graph to find the correct project config', async () => {
      await applicationGenerator(tree, {
        name: 'fancy-app',
      });
      await libraryGenerator(tree, {
        name: 'fancy-lib',
      });
      await componentGenerator(tree, {
        name: 'fancy-cmp',
        project: 'fancy-lib',
        export: true,
      });
      tree.write(
        'apps/fancy-app/src/app/blah.component.ts',
        `import {FancyCmpComponent} from '@something/fancy-lib'`
      );
      projectGraph = {
        nodes: {
          'fancy-app': {
            name: 'fancy-app',
            type: 'app',
            data: {
              ...readProjectConfiguration(tree, 'fancy-app'),
            },
          },
          'fancy-lib': {
            name: 'fancy-lib',
            type: 'lib',
            data: {
              ...readProjectConfiguration(tree, 'fancy-lib'),
            },
          },
        },
        dependencies: {
          'fancy-app': [
            {
              type: DependencyType.static,
              source: 'fancy-app',
              target: 'fancy-lib',
            },
          ],
        },
      };
      await cypressComponentConfiguration(tree, {
        project: 'fancy-lib',
        generateTests: false,
      });
      expect(
        readProjectConfiguration(tree, 'fancy-lib').targets['component-test']
      ).toEqual({
        executor: '@nrwl/cypress:cypress',
        options: {
          cypressConfig: 'libs/fancy-lib/cypress.config.ts',
          devServerTarget: 'fancy-app:build',
          skipServe: true,
          testingType: 'component',
        },
      });
    });
  });

  it('should work with simple components', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib',
    });

    await setup(tree, {
      project: 'my-lib',
      name: 'something',
      standalone: false,
    });
    projectGraph = {
      nodes: {
        something: {
          name: 'something',
          type: 'app',
          data: {
            ...readProjectConfiguration(tree, 'something'),
          },
        },
        'my-lib': {
          name: 'my-lib',
          type: 'lib',
          data: {
            ...readProjectConfiguration(tree, 'my-lib'),
          },
        },
      },
      dependencies: {
        'my-lib': [
          {
            type: DependencyType.static,
            source: 'my-lib',
            target: 'something',
          },
        ],
      },
    };
    await cypressComponentConfiguration(tree, {
      project: 'my-lib',
      buildTarget: 'something:build',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/my-lib/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with standalone component', async () => {
    await libraryGenerator(tree, {
      name: 'my-lib-standalone',
    });

    await setup(tree, {
      project: 'my-lib-standalone',
      name: 'something',
      standalone: true,
    });
    projectGraph = {
      nodes: {
        something: {
          name: 'something',
          type: 'app',
          data: {
            ...readProjectConfiguration(tree, 'something'),
          },
        },
        'my-lib-standalone': {
          name: 'my-lib-standalone',
          type: 'lib',
          data: {
            ...readProjectConfiguration(tree, 'my-lib-standalone'),
          },
        },
      },
      dependencies: {
        'my-lib-standalone': [
          {
            type: DependencyType.static,
            source: 'my-lib-standalone',
            target: 'something',
          },
        ],
      },
    };
    await cypressComponentConfiguration(tree, {
      project: 'my-lib-standalone',
      buildTarget: 'something:build',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/my-lib-standalone/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with complex component', async () => {
    await libraryGenerator(tree, {
      name: 'with-inputs-cmp',
    });

    await setup(tree, {
      project: 'with-inputs-cmp',
      name: 'something',
      standalone: false,
      withInputs: true,
      basePath: 'libs/with-inputs-cmp/src/lib',
    });
    projectGraph = {
      nodes: {
        something: {
          name: 'something',
          type: 'app',
          data: {
            ...readProjectConfiguration(tree, 'something'),
          },
        },
        'with-inputs-cmp': {
          name: 'with-inputs-cmp',
          type: 'lib',
          data: {
            ...readProjectConfiguration(tree, 'with-inputs-cmp'),
          },
        },
      },
      dependencies: {
        'with-inputs-cmp': [
          {
            type: DependencyType.static,
            source: 'with-inputs-cmp',
            target: 'something',
          },
        ],
      },
    };
    await cypressComponentConfiguration(tree, {
      project: 'with-inputs-cmp',
      buildTarget: 'something:build',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/with-inputs-cmp/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with complex standalone component', async () => {
    await libraryGenerator(tree, {
      name: 'with-inputs-standalone-cmp',
    });

    await setup(tree, {
      project: 'with-inputs-standalone-cmp',
      name: 'something',
      standalone: true,
      withInputs: true,
      basePath: 'libs/with-inputs-standalone-cmp/src/lib',
    });
    projectGraph = {
      nodes: {
        something: {
          name: 'something',
          type: 'app',
          data: {
            ...readProjectConfiguration(tree, 'something'),
          },
        },
        'with-inputs-standalone-cmp': {
          name: 'with-inputs-standalone-cmp',
          type: 'lib',
          data: {
            ...readProjectConfiguration(tree, 'with-inputs-standalone-cmp'),
          },
        },
      },
      dependencies: {
        'with-inputs-standalone-cmp': [
          {
            type: DependencyType.static,
            source: 'with-inputs-standalone-cmp',
            target: 'something',
          },
        ],
      },
    };
    await cypressComponentConfiguration(tree, {
      project: 'with-inputs-standalone-cmp',
      buildTarget: 'something:build',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      basePath: 'libs/with-inputs-standalone-cmp/src/lib',
      name: 'something',
    });
    expect(one.cy).toMatchSnapshot();
    expect(two.cy).toMatchSnapshot();
    expect(three.cy).toMatchSnapshot();
  });

  it('should work with secondary entry point libs', async () => {
    await applicationGenerator(tree, {
      name: 'my-cool-app',
    });
    await libraryGenerator(tree, {
      name: 'secondary',
      buildable: true,
    });
    await librarySecondaryEntryPointGenerator(tree, {
      name: 'button',
      library: 'secondary',
    });
    await componentGenerator(tree, {
      name: 'fancy-button',
      path: 'libs/secondary/src/lib/button',
      project: 'secondary',
      flat: true,
    });
    await componentGenerator(tree, {
      name: 'standalone-fancy-button',
      path: 'libs/secondary/src/lib/button',
      project: 'secondary',
      standalone: true,
      flat: true,
    });
    projectGraph = {
      nodes: {
        'my-cool-app': {
          name: 'my-cool-app',
          type: 'app',
          data: {
            ...readProjectConfiguration(tree, 'my-cool-app'),
          },
        },
        secondary: {
          name: 'secondary',
          type: 'lib',
          data: {
            ...readProjectConfiguration(tree, 'secondary'),
          },
        },
      },
      dependencies: {},
    };

    await cypressComponentConfiguration(tree, {
      generateTests: true,
      project: 'secondary',
      buildTarget: 'my-cool-app:build',
    });
    expect(
      tree.read(
        'libs/secondary/src/lib/button/fancy-button.component.cy.ts',
        'utf-8'
      )
    ).toMatchSnapshot();
    expect(
      tree.read(
        'libs/secondary/src/lib/button/standalone-fancy-button.component.cy.ts',
        'utf-8'
      )
    ).toMatchSnapshot();
  });

  it('should not overwrite existing component test', async () => {
    await libraryGenerator(tree, {
      name: 'cool-lib',
      flat: true,
    });
    await setup(tree, { project: 'cool-lib', name: 'abc', standalone: false });
    tree.write(
      'libs/cool-lib/src/lib/abc-one/abc-one.component.cy.ts',
      'should not overwrite abc-one'
    );
    tree.write(
      'libs/cool-lib/src/lib/abc-two/abc-two.component.cy.ts',
      'should not overwrite abc-two'
    );
    tree.write(
      'libs/cool-lib/src/lib/abc-three/abc-three.component.cy.ts',
      'should not overwrite abc-three'
    );
    projectGraph = {
      nodes: {
        abc: {
          name: 'abc',
          type: 'app',
          data: {
            ...readProjectConfiguration(tree, 'abc'),
          },
        },
        'cool-lib': {
          name: 'cool-lib',
          type: 'lib',
          data: {
            ...readProjectConfiguration(tree, 'cool-lib'),
          },
        },
      },
      dependencies: {},
    };
    await cypressComponentConfiguration(tree, {
      project: 'cool-lib',
      buildTarget: 'abc:build',
      generateTests: true,
    });

    const [one, two, three] = getCmpsFromTree(tree, {
      name: 'abc',
      basePath: 'libs/cool-lib/src/lib',
    });

    expect(one.cy).toEqual('should not overwrite abc-one');
    expect(two.cy).toEqual('should not overwrite abc-two');
    expect(three.cy).toEqual('should not overwrite abc-three');
  });

  // TODO: should we support this?
  it.skip('should handle multiple components per file', async () => {
    await libraryGenerator(tree, {
      name: 'multiple-components',
      flat: true,
    });

    await componentGenerator(tree, {
      name: 'cmp-one',
      project: 'multiple-components',
      flat: true,
    });
    await componentGenerator(tree, {
      name: 'cmp-two',
      project: 'multiple-components',
      flat: true,
    });
    tree.write(
      `libs/multiple-components/src/lib/cmp-one.component.ts`,
      `
import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'proj-cmp-one',
  templateUrl: './cmp-one.component.html',
  styleUrls: ['./cmp-one.component.css']
})
export class CmpOneComponent implements OnInit {

  constructor() { }

  ngOnInit(): void {
  }

}

@Component({
  selector: 'proj-cmp-one',
  template: '<h1>Hello World, {{abc}}</h1>',
  styles: []
})
export class CmpMultiComponent implements OnInit {
  @Input() name: string = 'abc'
  constructor() { }
  ngOnInit(): void {}
}
`
    );

    tree.write(
      '',
      `
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CmpOneComponent, CmpMultiComponent } from './cmp-one.component';
import { CmpTwoComponent } from './cmp-two.component';

@NgModule({
  imports: [
    CommonModule
  ],
  declarations: [
    CmpOneComponent,
    CmpTwoComponent
  ]
})
export class MultipleComponentsModule { }
`
    );

    await cypressComponentConfiguration(tree, {
      project: 'multiple-components',
      generateTests: true,
    });
    expect(
      tree.read(
        'libs/multiple-components/src/lib/cmp-one.component.cy.ts',
        'utf-8'
      )
    ).toEqual('');
  });
});

async function setup(
  tree: Tree,
  options: {
    name: string;
    project: string;
    standalone?: boolean;
    withInputs?: boolean;
    basePath?: string;
  }
) {
  await applicationGenerator(tree, {
    name: options.name,
    standalone: options.standalone,
  });
  for (const name of [
    `${options.name}-one`,
    `${options.name}-two`,
    `${options.name}-three`,
  ]) {
    await componentGenerator(tree, { project: options.project, name });

    if (options.withInputs) {
      const cmpPath = joinPathFragments(
        options.basePath,
        name,
        `${name}.component.ts`
      );
      const oldContent = tree.read(
        cmpPath,

        'utf-8'
      );

      const newContent = oldContent.replace(
        'constructor()',
        `
  @Input('buttonType') type = 'button';
  @Input() style: 'default' | 'fancy' | 'link' = 'default';
  @Input() age?: number;
  @Input() isOn = false;
  @Input() message: string | undefined;
  @Input() anotherProp: any;
  @Input() anotherNeverProp: never;

  constructor()`
      );

      tree.write(cmpPath, newContent);
    }
  }
}
function getCmpsFromTree(
  tree: Tree,
  options: { basePath: string; name: string }
) {
  return [
    `${options.name}-one`,
    `${options.name}-two`,
    `${options.name}-three`,
  ].map((n) => {
    expect(
      tree.exists(joinPathFragments(options.basePath, n, `${n}.component.ts`))
    ).toBeTruthy();
    expect(
      tree.exists(
        joinPathFragments(options.basePath, n, `${n}.component.cy.ts`)
      )
    ).toBeTruthy();
    return {
      cmp: tree.read(
        joinPathFragments(options.basePath, n, `${n}.component.ts`),
        'utf-8'
      ),
      cy: tree.read(
        joinPathFragments(options.basePath, n, `${n}.component.cy.ts`),
        'utf-8'
      ),
    };
  });
}
