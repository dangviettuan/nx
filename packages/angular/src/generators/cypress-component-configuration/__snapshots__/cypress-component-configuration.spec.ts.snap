// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cypress Component Testing Configuration should work with complex component 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingOneComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex component 2`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingTwoComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex component 3`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingThreeComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex standalone component 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingOneComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex standalone component 2`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingTwoComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with complex standalone component 3`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingThreeComponent, {
           ...config,
           componentProperties: {
               type:  'button',
               style:  'default',
               age:  0,
               isOn:  false,
          }
       });
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with secondary entry point libs 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { FancyButtonComponent } from './fancy-button.component';

describe(FancyButtonComponent.name, () => {
  const config: MountConfig<FancyButtonComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(FancyButtonComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with secondary entry point libs 2`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { StandaloneFancyButtonComponent } from './standalone-fancy-button.component';

describe(StandaloneFancyButtonComponent.name, () => {
  const config: MountConfig<StandaloneFancyButtonComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(StandaloneFancyButtonComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with simple components 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingOneComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with simple components 2`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingTwoComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with simple components 3`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingThreeComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with standalone component 1`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingOneComponent } from './something-one.component';

describe(SomethingOneComponent.name, () => {
  const config: MountConfig<SomethingOneComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingOneComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with standalone component 2`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingTwoComponent } from './something-two.component';

describe(SomethingTwoComponent.name, () => {
  const config: MountConfig<SomethingTwoComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingTwoComponent, config);
  })
})
"
`;

exports[`Cypress Component Testing Configuration should work with standalone component 3`] = `
"import { MountConfig, mount } from 'cypress/angular';
import { SomethingThreeComponent } from './something-three.component';

describe(SomethingThreeComponent.name, () => {
  const config: MountConfig<SomethingThreeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(SomethingThreeComponent, config);
  })
})
"
`;
